File: ps/letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        unordered_map<char, string> letters;
        
        letters['2'] = "abc";
        letters['3'] = "def";
        letters['4'] = "ghi";
        letters['5'] = "jkl";
        letters['6'] = "mno";
        letters['7'] = "pqrs";
        letters['8'] = "tuv";
        letters['9'] = "wxyz";
        
        vector<string> res;
        if (digits.size() == 0) return res;
        enumerator(res, letters, digits, 0, "");
        return res;
    }
    
    void enumerator(vector<string> &res, unordered_map<char, string> letters, string digits, int idx, string alphs) {
        int i;
        if (digits.size() == idx) {
            res.push_back(alphs);
            return;
        }
        for (i = 0 ; i<letters[digits[idx]].size();i++) {
            enumerator(res, letters, digits, idx + 1, alphs + letters[digits[idx]][i]);
        }
        return;
    }
};

File: ps/letter-combinations-of-a-phone-number.py
class Solution:
    dialDict = {'2': ['a', 'b', 'c'],
                '3': ['d', 'e', 'f'],
                '4': ['g', 'h', 'i'],
                '5': ['j', 'k', 'l'],
                '6': ['m', 'n', 'o'],
                '7': ['p', 'q', 'r', 's'],
                '8': ['t', 'u', 'v'],
                '9': ['w', 'x', 'y', 'z']}
    res = []

    def letterCombinations(self, digits: str) -> List[str]:
        if len(digits) == 0:
            return []
        self.res = []
        self.enumerates(digits, 0, "")
        return self.res

    def enumerates(self, digits, index, combs):
        if index == len(digits):
            self.res.append(combs)
            return
        digit = digits[index]
        for char in self.dialDict[digit]:
            self.enumerates(digits, index + 1, combs + char)